### PR TITLE
Added support for 'extract base class'

### PR DESCRIPTION
**Pull request by filipw**
This is a new feature in Roslyn, just added this August.

It doesn't light up for us because it requires UI option interaction. This PR adds some reasonable defaults and a dispatch proxy to implement the internal interface - following similar model to what we have in e.g. extract interface already.

extractclass

Note that the name of the refactoring will change when we pull in new Roslyn build. It's currently `Pull member(s) up to new base class...` but in the new build it's already renamed to  `Extract base class...`

### CODE REVIEW

File: src/OmniSharp.Roslyn/WorkspaceServices/ExtractClassWorkspaceService.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    method,
+                    parameters = method.GetParameters(),
+                    genericArguments = method.GetGenericArguments()
+                })
+                .Where(method => method.genericArguments.Length == 1 && method.parameters.Length == 1)
+                .Select(x => x.method)
+                .First()
+                .MakeGenericMethod(extractClassMemberAnalysisResultType).Invoke(null, new object[]
+                {
+                    // at this point we have IEnumerable<object> and need to cas to IEnumerable<ExtractClassMemberAnalysisResult>
```
Comment (JoeRobich) [inline src/OmniSharp.Roslyn/WorkspaceServices/ExtractClassWorkspaceService.cs:53]:
```suggestion
                    // at this point we have IEnumerable<object> and need to cast to IEnumerable<ExtractClassMemberAnalysisResult>
```

File: src/OmniSharp.Roslyn/WorkspaceServices/ExtractClassWorkspaceService.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                    // which we can then pass to ImmutableArray.CreateRange<ExtractClassMemberAnalysisResult>()
+                    genericCast.Invoke(null, new object[]
+                    {
+                        // this constrcutor corresponds to
```
Comment (JoeRobich) [inline src/OmniSharp.Roslyn/WorkspaceServices/ExtractClassWorkspaceService.cs:57]:
```suggestion
                        // this constructor corresponds to
```

File: src/OmniSharp.Roslyn/WorkspaceServices/ExtractClassWorkspaceService.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                        {
+                            symbol,
+                            false
+                        }))
+                    })
+                });
+
+            const string name = "NewBaseType";
+
+            // this constrcutor corresponds to
```
Comment (JoeRobich) [inline src/OmniSharp.Roslyn/WorkspaceServices/ExtractClassWorkspaceService.cs:71]:
```suggestion
            // this constructor corresponds to
```

File: tests/OmniSharp.Roslyn.CSharp.Tests/CodeActionsWithOptionsFacts.cs [hunk 4]
```diff
@@ -15,6 +15,63 @@ public CodeActionsWithOptionsFacts(ITestOutputHelper output)
...
+                    public string Property { get; set; }
+
+                    public void Method() { }
+                }
+
+                public class Class1 : NewBaseType
+                {
+                }";
+
+            // TODO: this refactoring was renamed to 'Extract base class...' in latest Roslyn
```
Comment (JoeRobich) [inline tests/OmniSharp.Roslyn.CSharp.Tests/CodeActionsWithOptionsFacts.cs:40]: Thanks for documenting this!

**Review by JoeRobich — APPROVED** [review]

File: tests/OmniSharp.Roslyn.CSharp.Tests/CodeActionsWithOptionsFacts.cs (hunk 4 above)
Comment (filipw) [reply tests/OmniSharp.Roslyn.CSharp.Tests/CodeActionsWithOptionsFacts.cs:40]: for the next lucky winner that updates Roslyn 😀

**Pull request merged**